Remove progressbar from OffenseCountFormatter

Kjarrigan opened the pull request:
Recently I started using multiple outputs to speed up my workflow especially for older projects which didn't use rubocop before. Looking through the docs and the code it felt odd, that no formatter besides the **OffenseCountFormatter** uses a progressbar.

So the PR removes the progressbar to be more consistent. If I want a progressbar I can use **FuubarStyleFormatter**

```bash
$ rake default
Files:         514
Modules:        80 (   11 undocumented)
Classes:       485 (    3 undocumented)
Constants:     751 (  738 undocumented)
Attributes:     26 (    0 undocumented)
Methods:      1163 ( 1037 undocumented)
 28.58% documented
* generated /home/holger.arndt/Projects/rubocop/manual/cops_bundler.md
* generated /home/holger.arndt/Projects/rubocop/manual/cops_gemspec.md
* generated /home/holger.arndt/Projects/rubocop/manual/cops_layout.md
* generated /home/holger.arndt/Projects/rubocop/manual/cops_lint.md
* generated /home/holger.arndt/Projects/rubocop/manual/cops_metrics.md
* generated /home/holger.arndt/Projects/rubocop/manual/cops_naming.md
* generated /home/holger.arndt/Projects/rubocop/manual/cops_performance.md
* generated /home/holger.arndt/Projects/rubocop/manual/cops_rails.md
* generated /home/holger.arndt/Projects/rubocop/manual/cops_security.md
* generated /home/holger.arndt/Projects/rubocop/manual/cops_style.md
Starting test-queue master (/tmp/test_queue_26493_47227101006560.sock)

==> Summary (8 workers in 9.4735s)

    [ 1]                                    3297 examples, 0 failures       136 suites in 8.1272s      (pid 27015 exit 0 )
    [ 2]                         6109 examples, 0 failures, 5 pending        32 suites in 9.4596s      (pid 27018 exit 0 )
    [ 3]                                    3317 examples, 0 failures        57 suites in 9.4564s      (pid 27021 exit 0 )
    [ 4]                         1716 examples, 0 failures, 1 pending        48 suites in 9.4545s      (pid 27024 exit 0 )
    [ 5]                                    3546 examples, 0 failures        96 suites in 9.4531s      (pid 27027 exit 0 )
    [ 6]                                     481 examples, 0 failures        13 suites in 9.4520s      (pid 27030 exit 0 )
    [ 7]                         3333 examples, 0 failures, 1 pending       120 suites in 9.4389s      (pid 27033 exit 0 )
    [ 8]                                     948 examples, 0 failures        21 suites in 9.4122s      (pid 27036 exit 0 )
```

WDYT?

bbatsov commented:
I'm fine with removing this as I don't really use it. One less dependency for the project is always an improvement for me. Not sure about the end users and the other people on @rubocop-hq/rubocop-core.

jonas054 commented:
@bbatsov We still depend on `ruby-progressbar` because of `FuubarStyleFormatter`.

@Kjarrigan said:

> If I want a progressbar I can use FuubarStyleFormatter

I'd like to see a formatter that adds a progress bar and nothing more. The `FuubarStyleFormatter` also adds reporting of individual offenses with the highlighted source lines. That's not suitable for combining with other formatters. Could you add a formatter like that in this PR? Otherwise I fear we're removing something that some people find useful.

bquorning commented:
Especially when working on large code bases, I like to have some indication of progress. I think that for formatters that don’t output anything *per file* (implementing `#file_started` and/or `#file_finished`), I would prefer having a progress bar as well.

Perhaps we could add a config flag to enable/disable the progress bar (like e.g. `curl` has `--progress-bar` and `--silent`), and make a conditional implementation in `RuboCop::Formatter::BaseFormatter#started` and `RuboCop::Formatter::BaseFormatter#file_finished`?

Kjarrigan commented:
> Could you add a formatter like that in this PR? Otherwise I fear we're removing something that some people find useful.

I certainly can. Will look into this.

> Perhaps we could add a config flag to enable/disable the progress bar
I thought about something like this too, but didn't want to dive to deep into argument parsing etc "just" to make it a bit more consistent.

Combined with @jonas054 approach to make a progressbar-only formatter.` --progressbar` could be a _"shortcut"_ to enable the progressbar only formatter? Or should we bother injecting progressbar in all the things. How should the output look like if I activate 3 formatters and `--progressbar` how to prevent seeing 3 progressbars?

Kjarrigan commented:
Ok. Rebased and added a `ProgressbarFormatter` which is basically the code I removed in the first commit. Also added the needed connections to the CLI, FormatterSet, etc but there I got some _problems_ with the naming.

At first I wanted to name it `progressbar` but this collides with the logic how the formatters are picked (`key.starts_with?`) so effectively the `progress` formatter would no longer work. The I tried `onlyprogressbar` which conflicts with the `[o]ffense` formatter so that we would break it for all user who use `-f o` at the moment (the new _shortcut_ would be `-f of`

So for now I made it simply `[b]ar` which isn't that expressive but doesn't interfere with any other formatter.

Kjarrigan commented:
Rebased. Any thoughts on my last comment regarding the naming? Any other things to get this resolved?(either merged or closed)

jonas054 commented:
The naming (`bar`) is good IMO. I would like this to be merged, but you have one remaining violation. Please take a look.

bbatsov commented:
I'm closing this PR due to no recent activity. Feel free to re-open it if you ever come back to it.
